projects/RabbitMQ.Client/client/impl/Connection.Receive.cs

```diff
@@ -66,57 +66,68 @@ private void MainLoop()
             FinishClose();
         }
 
-        private void ReceiveLoop()
+        private async Task ReceiveLoop()
         {
             while (!_closed)
             {
-                InboundFrame frame = _frameHandler.ReadFrame();
+                while (_frameHandler.TryReadFrame(out InboundFrame frame))
+                {
+                    NotifyHeartbeatListener();
+                    ProcessFrame(frame);
+                }
+
+                // Done reading frames synchronously, go async
+                InboundFrame asyncFrame = await _frameHandler.ReadFrameAsync().ConfigureAwait(false);
                 NotifyHeartbeatListener();
+                ProcessFrame(asyncFrame);
+            }
+        }
 
-                bool shallReturn = true;
-                if (frame.Channel == 0)
+        private void ProcessFrame(InboundFrame frame)
+        {
+            bool shallReturn = true;
+            if (frame.Channel == 0)
+            {
+                if (frame.Type == FrameType.FrameHeartbeat)
                 {
-                    if (frame.Type == FrameType.FrameHeartbeat)
-                    {
-                        // Ignore it: we've already just reset the heartbeat
-                    }
-                    else
-                    {
-                        // In theory, we could get non-connection.close-ok
-                        // frames here while we're quiescing (m_closeReason !=
-                        // null). In practice, there's a limited number of
-                        // things the server can ask of us on channel 0 -
-                        // essentially, just connection.close. That, combined
-                        // with the restrictions on pipelining, mean that
-                        // we're OK here to handle channel 0 traffic in a
-                        // quiescing situation, even though technically we
-                        // should be ignoring everything except
-                        // connection.close-ok.
-                        shallReturn = _session0.HandleFrame(in frame);
-                    }
+                    // Ignore it: we've already just reset the heartbeat
```

Comment:
I don't know but something with the just reads weird to me.

```suggestion
                    // Ignore it: we've already recently reset the heartbeat
```

or 

```suggestion
                    // Ignore it: The heartbeat was recently reset
```